crates/uv-git/src/git.rs

```diff
@@ -133,24 +124,69 @@ pub(crate) struct GitRemote {
 /// A local clone of a remote repository's database. Multiple [`GitCheckout`]s
 /// can be cloned from a single [`GitDatabase`].
 pub(crate) struct GitDatabase {
-    /// The remote repository where this database is fetched from.
-    remote: GitRemote,
-    /// Path to the root of the underlying Git repository on the local filesystem.
-    path: PathBuf,
     /// Underlying Git repository instance for this database.
-    repo: git2::Repository,
+    repo: GitRepository,
 }
 
-/// A local checkout of a particular revision from a [`GitDatabase`].
-pub(crate) struct GitCheckout<'a> {
-    /// The git database where this checkout is cloned from.
-    database: &'a GitDatabase,
-    /// Path to the root of the underlying Git repository on the local filesystem.
-    path: PathBuf,
+/// A local checkout of a particular revision from a [`GitRepository`].
+pub(crate) struct GitCheckout {
     /// The git revision this checkout is for.
-    revision: git2::Oid,
+    revision: GitOid,
     /// Underlying Git repository instance for this checkout.
-    repo: git2::Repository,
+    repo: GitRepository,
+}
+
+/// A local Git repository.
+pub(crate) struct GitRepository {
+    /// Path to the underlying Git repository on the local filesystem.
+    path: PathBuf,
+}
+
+impl GitRepository {
+    /// Opens an existing Git repository at `path`.
+    pub(crate) fn open(path: &Path) -> Result<GitRepository> {
+        // Make sure there is a Git repository at the specified path.
+        ProcessBuilder::new("git")
+            .arg("rev-parse")
+            .cwd(path)
+            .exec_with_streaming(&mut silent, &mut silent, true)?;
+
+        Ok(GitRepository {
+            path: path.to_path_buf(),
+        })
+    }
+
+    /// Initializes a Git repository at `path`.
+    fn init(path: &Path) -> Result<GitRepository> {
+        // TODO: see if this still necessary
+        // Skip anything related to templates, they just call all sorts of issues as
+        // we really don't want to use them yet they insist on being used. See #6240
+        // for an example issue that comes up.
+        // opts.external_template(false);
+
+        // Initialize the repository.
+        ProcessBuilder::new("git")
+            .arg("init")
+            .cwd(path)
+            .exec_with_streaming(&mut silent, &mut silent, true)?;
```

Comment:
Is there any risk of dropping important output here? (What did we do with Git CLI output before, when fetching?)

Reply:
This is the same thing we've always done in `fetch_with_cli`. The `true` means that stderr output will still be captured in the error message if the command fails, the `silent` callbacks just make sure that the stdout/stderr output is not piped to the user's terminal (`Command::{stdout,stderr}` are overridden).

Reply:
Turns out we can use `exec_with_output` here instead, which is a bit clearer.